src/box/xrow.c

```diff
@@ -1705,70 +1705,79 @@ xrow_decode_ballot(const struct xrow_header *row, struct ballot *ballot)
+ * the iproto keys are fixed.
+ */
+struct replication_request {
+	/** IPROTO_CLUSTER_UUID. */
+	struct tt_uuid *replicaset_uuid;
+	/** IPROTO_INSTANCE_UUID. */
+	struct tt_uuid *instance_uuid;
+	/** IPROTO_VCLOCK. */
+	struct vclock *vclock;
+	/** IPROTO_ID_FILTER. */
+	uint32_t *id_filter;
+	/** IPROTO_SERVER_VERSION. */
+	uint32_t *version_id;
+	/** IPROTO_REPLICA_ANON. */
+	bool *is_anon;
+};
 
-int
-xrow_encode_subscribe(struct xrow_header *row,
-		      const struct tt_uuid *replicaset_uuid,
-		      const struct tt_uuid *instance_uuid,
-		      const struct vclock *vclock, bool anon,
-		      uint32_t id_filter)
+/** Encode a replication request template. */
+static int
+xrow_encode_replication_request(struct xrow_header *row,
+				const struct replication_request *req,
+				uint16_t type)
 {
 	memset(row, 0, sizeof(*row));
-	size_t size = XROW_BODY_LEN_MAX +
-		      mp_sizeof_vclock_ignore0(vclock);
+	size_t size = XROW_BODY_LEN_MAX;
+	if (req->vclock != NULL)
+		size += mp_sizeof_vclock_ignore0(req->vclock);
 	char *buf = (char *) region_alloc(&fiber()->gc, size);
 	if (buf == NULL) {
 		diag_set(OutOfMemory, size, "region_alloc", "buf");
 		return -1;
 	}
+	uint32_t id_filter = req->id_filter != NULL ? *req->id_filter : 0;
+	uint32_t map_size = 0;
+	map_size += req->replicaset_uuid != NULL;
+	map_size += req->instance_uuid != NULL;
+	map_size += req->vclock != NULL;
+	map_size += id_filter != 0;
+	map_size += req->is_anon != NULL;
+	map_size += req->version_id != NULL;
+
 	char *data = buf;
-	int filter_size = bit_count_u32(id_filter);
-	data = mp_encode_map(data, filter_size != 0 ? 6 : 5);
-	data = mp_encode_uint(data, IPROTO_CLUSTER_UUID);
-	data = xrow_encode_uuid(data, replicaset_uuid);
-	data = mp_encode_uint(data, IPROTO_INSTANCE_UUID);
-	data = xrow_encode_uuid(data, instance_uuid);
-	data = mp_encode_uint(data, IPROTO_VCLOCK);
-	data = mp_encode_vclock_ignore0(data, vclock);
-	data = mp_encode_uint(data, IPROTO_SERVER_VERSION);
-	data = mp_encode_uint(data, tarantool_version_id());
-	data = mp_encode_uint(data, IPROTO_REPLICA_ANON);
-	data = mp_encode_bool(data, anon);
-	if (filter_size != 0) {
+	data = mp_encode_map(data, map_size);
+	if (req->replicaset_uuid != NULL) {
+		data = mp_encode_uint(data, IPROTO_CLUSTER_UUID);
+		data = xrow_encode_uuid(data, req->replicaset_uuid);
+	}
+	if (req->instance_uuid != NULL) {
+		data = mp_encode_uint(data, IPROTO_INSTANCE_UUID);
+		data = xrow_encode_uuid(data, req->instance_uuid);
+	}
+	if (req->vclock != NULL) {
+		data = mp_encode_uint(data, IPROTO_VCLOCK);
+		data = mp_encode_vclock_ignore0(data, req->vclock);
+	}
+	if (req->version_id != NULL) {
+		data = mp_encode_uint(data, IPROTO_SERVER_VERSION);
+		data = mp_encode_uint(data, *req->version_id);
+	}
+	if (req->is_anon != NULL) {
+		data = mp_encode_uint(data, IPROTO_REPLICA_ANON);
+		data = mp_encode_bool(data, *req->is_anon);
+	}
+	if (id_filter != 0) {
```

Comment:
Let's encode it unconditionally (when `req->id_filter` isn't NULL). This'd be consistent with other fields. I don't see why we are trying to save a couple of bytes on a subscribe request.

Reply:
The intention was not to change the behaviour - previously we didn't encode the filter when it was 0. But I don't mind to alter it. Done.

Reply:
Thanks